Buffer/scullBuffer.c

```diff
@@ -92,139 +96,183 @@ static void scull_setup_cdev(struct scull_buffer *dev)
-		dev->writerCnt++;
-		
-	up(&dev->sem);
-	return 0;
+  struct scull_buffer *dev;
+  /* get and store the container scull_buffer */
+  dev = container_of(inode->i_cdev, struct scull_buffer, cdev);
+  filp->private_data = dev;
+
+  if (down_interruptible(&dev->sem))
+    return -ERESTARTSYS;
+
+  if (filp->f_mode & FMODE_READ)
+    dev->readerCnt++;
+  if (filp->f_mode & FMODE_WRITE)
+    dev->writerCnt++;
+
+  up(&dev->sem);
+  return 0;
 }
 
 /* 
  * Implementation of the close system call
-*/
+ */
 int scullBuffer_release(struct inode *inode, struct file *filp)
 {
-	struct scull_buffer *dev = (struct scull_buffer *)filp->private_data;
-	if (down_interruptible(&dev->sem) )
-		return -ERESTARTSYS;
-		
-	if (filp->f_mode & FMODE_READ)
-		dev->readerCnt--;
-	if (filp->f_mode & FMODE_WRITE)
-		dev->writerCnt--;
-		
-	up(&dev->sem);
-	return 0;
+  struct scull_buffer *dev = (struct scull_buffer *)filp->private_data;
+  if (down_interruptible(&dev->sem) )
+    return -ERESTARTSYS;
+
+  if (filp->f_mode & FMODE_READ)
+    dev->readerCnt--;
+  if (filp->f_mode & FMODE_WRITE)
+    dev->writerCnt--;
+
+  up(&dev->sem);
+  return 0;
 }
 
 /*
  * Implementation of the read system call
-*/
-ssize_t scullBuffer_read(
-	struct file *filp,
-	char __user *buf,
-	size_t count,
-	loff_t *f_pos)
+ */
+ssize_t scullBuffer_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
 {
-	struct scull_buffer *dev = (struct scull_buffer *)filp->private_data;
-	ssize_t countRead = 0;
-	
-	/* get exclusive access */
-	if (down_interruptible(&dev->sem))
-		return -ERESTARTSYS;	
-		
-	printk(KERN_DEBUG "scullBuffer: read called count= %d\n", count);
-	printk(KERN_DEBUG "scullBuffer: cur pos= %lld, size= %d \n", *f_pos, dev->size);
-	
-	/* have we crossed reading the device? */
-	if( *f_pos >= dev->size)
-		goto out;
-	/* read till the end of device */
-	if( (*f_pos + count) > dev->size)
-		count = dev->size - *f_pos;
-	
-	printk(KERN_DEBUG "scullBuffer: reading %d bytes\n", (int)count);
-	/* copy data to user space buffer */
-	if (copy_to_user(buf, dev->bufferPtr + *f_pos, count)) {
-		countRead = -EFAULT;
-		goto out;
-	}
-	printk(KERN_DEBUG "scullBuffer: new pos= %lld\n", *f_pos);
-	*f_pos += count;
-	countRead = count;	
-	
-	/* now we're done release the semaphore */
-	out: 
-	up(&dev->sem);
-	return countRead;
+  struct scull_buffer *dev = (struct scull_buffer *)filp->private_data;
+  int currFront = 0;
+  ssize_t countRead = 0;
+  int len = 0;
+  int res = 0;
+
+  /* Atomically check if empty */
+  if (down_interruptible(&dev->sem))
+    return -ERESTARTSYS;
+
+  /* Check if the buffer is empty */
+  if(dev->front >= dev->nextEmpty && dev->writerCnt <= 0) {
+    printk(KERN_DEBUG "Buffer is empty\n");
+    up(&dev->sem);
+    return 0;
+  }
+  currFront = dev->front;
+  /* Move the front of the buffer up */
+  dev->front += 516;
+  up(&dev->sem);
+
+  /* Get mutex for accessing buffer */
+  if (down_interruptible(&dev->full))
+    return -ERESTARTSYS;
```

Comment:
If there are producers running, the read will block when it tries to obtain this semaphore